ads/google/csa.js

```diff
@@ -0,0 +1,327 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {validateData, loadScript} from '../../3p/3p';
+import {tryParseJson} from '../../src/json.js';
+
+// Keep track of current height of AMP iframe
+let currentAmpHeight = null;
+
+/**
+ * Request Custom Search Ads (Adsense for Search or AdSense for Shopping).
+ * @param {!Window} global The window object of the iframe
+ * @param {!Object} data
+ */
+export function csa(global, data) {
+
+  // Get parent width in case we want to override
+  const width = global.document.body.clientWidth;
+
+  validateData(data, [], [
+    'afshPageOptions',
+    'afshAdblockOptions',
+    'afsPageOptions',
+    'afsAdblockOptions',
+    'ampSlotIndex',
+  ]);
+
+  // Add the ad container to the document
+  const csaContainerDiv = global.document.createElement('div');
+  const containerId = 'csacontainer';
+  csaContainerDiv.id = containerId;
+  global.document.body.appendChild(csaContainerDiv);
+
+  // Parse AFSh page options
+  const afshPageOptions = Object.assign(
+    Object(tryParseJson(data['afshPageOptions'])),
+    {source: 'amp', referer: global.context.referrer});
+
+
+  // Parse AFSh adblock options
+  const afshAdblockOptions = Object.assign(
+    Object(tryParseJson(data['afshAdblockOptions'])),
+    {container: containerId});
+
+  // Set the width to the width of the screen if necessary
+  if (afshAdblockOptions != null &&
+    afshAdblockOptions['width'] == 'auto') {
+    afshAdblockOptions['width'] = width;
+  }
+
+  // Parse AFS page options
+  const afsPageOptions = Object.assign(
+    Object(tryParseJson(data['afsPageOptions'])),
+    {source: 'amp', referer: global.context.referrer});
+
+  // Parse AFS adblock options
+  const afsAdblockOptions = Object.assign(
+    Object(tryParseJson(data['afsAdblockOptions'])),
+    {container: containerId});
+
+  /* Time in ms to wait before executing orientation change function logic */
+  const orientationChangeTimeout = 250;
+
+  /**
+   * Resize the AMP iframe if the CSA container changes in size upon rotation.
+   * This is needed for an iOS bug found in versions 10.0.1 and below that
+   * doesn't properly reflow the iframe upon orientation change.
+   */
+  global.addEventListener('orientationchange', () => {
+
+    // Save the height of the container before the event listener triggers
+    const oldHeight =
+      global.document.getElementById('csacontainer').style.height;
+
+    setTimeout(() => {
+
+      // Force DOM reflow and repaint
+      /*eslint-disable no-unused-vars*/
+      const ignore = global.document.body.offsetHeight;
```

Comment:
Also this section was causing me compiler errors when I tried to deploy to heroku (#5351) so this code was added to fix that.